File: Source/Swift/MDSBatchInfo.swift
//
//  MDSBatchInfo.swift
//  Mini Document Storage
//
//  Created by Stevo on 10/9/19.
//  Copyright © 2019 Stevo Brock. All rights reserved.
//

import Foundation

//----------------------------------------------------------------------------------------------------------------------
// MARK: MDSBatchInfo
class MDSBatchInfo<T> {

	// MARK: AddAttachmentInfo
	struct AddAttachmentInfo {

		// MARK: Properties
		let	attachmentID :String
		let	revision :Int
		let	info :[String : Any]
		let	content :Data

		var	attachmentInfo :MDSDocument.AttachmentInfo
				{ MDSDocument.AttachmentInfo(id: self.attachmentID, revision: self.revision, info: self.info) }

		// MARK: Lifecycle methods
		//--------------------------------------------------------------------------------------------------------------
		init(info :[String : Any], content :Data) {
			// Setup
			self.attachmentID = UUID().base64EncodedString
			self.revision = 1

			// Store
			self.info = info
			self.content = content
		}
	}

	// MARK: UpdateAttachmentInfo
	struct UpdateAttachmentInfo {

		// MARK: Properties
		let	attachmentID :String
		let	currentRevision :Int
		let	info :[String : Any]
		let	content :Data

		var	attachmentInfo :MDSDocument.AttachmentInfo
				{ MDSDocument.AttachmentInfo(id: self.attachmentID, revision: self.currentRevision, info: self.info) }
	}

	// MARK: RemoveAttachmentInfo
	struct RemoveAttachmentInfo {

		// MARK: Properties
		let	attachmentID :String
	}

	// MARK: DocumentInfo
	class DocumentInfo<T> {

		// MARK: Procs
		typealias ValueProc = (_ property :String) -> Any?

		// MARK: Properties
						let	documentType :String
						let	documentBacking :T?
						let	creationDate :Date

		private(set)	var	updatedPropertyMap :[String : Any]?
		private(set)	var	removedProperties :Set<String>?
		private(set)	var	modificationDate :Date
		private(set)	var	addAttachmentInfos = [AddAttachmentInfo]()
		private(set)	var	updateAttachmentInfos = [UpdateAttachmentInfo]()
		private(set)	var	removeAttachmentInfos = [RemoveAttachmentInfo]()
		private(set)	var	removed = false

		private			let	initialPropertyMap :[String : Any]?

		// MARK: Lifecycle methods
		//--------------------------------------------------------------------------------------------------------------
		fileprivate init(documentType :String, documentBacking :T?, creationDate :Date, modificationDate :Date,
				initialPropertyMap :[String : Any]?) {
			// Store
			self.documentType = documentType
			self.documentBacking = documentBacking
			self.creationDate = creationDate

			self.modificationDate = modificationDate

			self.initialPropertyMap = initialPropertyMap
		}

		// MARK: Instance methods
		//--------------------------------------------------------------------------------------------------------------
		func value(for property :String) -> Any? {
			// Check for document removed
			if self.removed {
				// Document removed
				return nil
			} else if self.removedProperties?.contains(property) ?? false {
				// Removed
				return nil
			} else {
				// Not removed
				return self.updatedPropertyMap?[property] ?? self.initialPropertyMap?[property]
			}
		}

		//--------------------------------------------------------------------------------------------------------------
		func set(_ value :Any?, for property :String) {
			// Check if have value
			if value != nil {
				// Have value
				if self.updatedPropertyMap != nil {
					// Have updated info
					self.updatedPropertyMap![property] = value
				} else {
					// First updated info
					self.updatedPropertyMap = [property : value!]
				}

				self.removedProperties?.remove(property)
			} else {
				// Removing value
				self.updatedPropertyMap?[property] = nil

				if self.removedProperties != nil {
					// Have removed properties
					self.removedProperties!.insert(property)
				} else {
					// First removed property
					self.removedProperties = Set<String>([property])
				}
			}

			// Modified
			self.modificationDate = Date()
		}

		//--------------------------------------------------------------------------------------------------------------
		func remove() { self.removed = true; self.modificationDate = Date() }

		//--------------------------------------------------------------------------------------------------------------
		func attachmentInfoMap(applyingChangesTo attachmentInfoMap :MDSDocument.AttachmentInfoMap) ->
				MDSDocument.AttachmentInfoMap {
			// Start with initial
			var	updatedAttachmentInfoMap = attachmentInfoMap

			// Process adds
			self.addAttachmentInfos.forEach()
					{ updatedAttachmentInfoMap[$0.attachmentID] = $0.attachmentInfo }

			// Process updates
			self.updateAttachmentInfos.forEach()
					{ updatedAttachmentInfoMap[$0.attachmentID] = $0.attachmentInfo }

			// Process removes
			updatedAttachmentInfoMap.removeValues(
					forKeys: self.removeAttachmentInfos.map({ $0.attachmentID }))

			return updatedAttachmentInfoMap
		}

		//--------------------------------------------------------------------------------------------------------------
		func attachmentContent(for attachmentID :String) -> Data? {
			// Check if have info on attachment
			if let addAttachmentInfo = self.addAttachmentInfos.first(where: { $0.attachmentID == attachmentID}) {
				// Have add
				return addAttachmentInfo.content
			} else if let updateAttachmentInfo =
					self.updateAttachmentInfos.first(where: { $0.attachmentID == attachmentID} ) {
				// Have update
				return updateAttachmentInfo.content
			} else {
				// No valid attachment
				return nil
			}
		}

		//--------------------------------------------------------------------------------------------------------------
		func attachmentAdd(info :[String : Any], content :Data) -> MDSDocument.AttachmentInfo {
			// Setup
			let	addAttachmentInfo = AddAttachmentInfo(info: info, content: content)

			// Add info
			self.addAttachmentInfos.append(addAttachmentInfo)

			// Modified
			self.modificationDate = Date()

			return addAttachmentInfo.attachmentInfo
		}

		//--------------------------------------------------------------------------------------------------------------
		func attachmentUpdate(attachmentID :String, currentRevision :Int, info :[String : Any], content :Data) {
			// Add info
			self.updateAttachmentInfos.append(
					UpdateAttachmentInfo(attachmentID: attachmentID, currentRevision: currentRevision, info: info,
							content: content))

			// Modified
			self.modificationDate = Date()
		}

		//--------------------------------------------------------------------------------------------------------------
		func attachmentRemove(attachmentID :String) {
			// Add info
			self.removeAttachmentInfos.append(RemoveAttachmentInfo(attachmentID: attachmentID))

			// Modified
			self.modificationDate = Date()
		}
	}

	// MARK: Procs
	typealias AddAttachmentProc =
				(_ documentType :String, _ documentID :String, _ documentBacking :T?,
						_ addAttachmentInfo :AddAttachmentInfo) -> Void
	typealias UpdateAttachmentProc =
				(_ documentType :String, _ documentID :String, _ documentBacking :T?,
						_ updateAttachmentInfo :UpdateAttachmentInfo) -> Void
	typealias RemoveAttachmentProc =
				(_ documentType :String, _ documentID :String, _ documentBacking :T?,
						_ removeAttachmentInfo :RemoveAttachmentInfo) -> Void

	// MARK: Properties
	private	var	documentInfoMap = [/* Document ID */ String : DocumentInfo<T>]()
	private	let	documentInfoMapLock = ReadPreferringReadWriteLock()

	private	var	associationUpdatesByAssociationName = [/* Name */ String : [MDSAssociation.Update]]()

	// MARK: Instance methods
	//------------------------------------------------------------------------------------------------------------------
	func associationNoteUpdated(for name :String, updates :[MDSAssociation.Update]) {
		// Add
		self.associationUpdatesByAssociationName.appendArrayValueElements(key: name, values: updates)
	}

	//------------------------------------------------------------------------------------------------------------------
	func associationIterateChanges(proc :(_ name :String, _ updates :[MDSAssociation.Update]) -> Void) {
		// Iterate info
		self.associationUpdatesByAssociationName.forEach() { proc($0.key, $0.value) }
	}

	//------------------------------------------------------------------------------------------------------------------
	func associationGetChanges(for name :String) -> [MDSAssociation.Update]? {
		// Return updates
		return self.associationUpdatesByAssociationName[name]
	}

	//------------------------------------------------------------------------------------------------------------------
	func documentAdd(documentType :String, documentID :String, documentBacking :T? = nil, creationDate :Date,
			modificationDate :Date, initialPropertyMap :[String : Any]? = nil) -> DocumentInfo<T> {
		// Setup
		let	documentInfo =
					DocumentInfo(documentType: documentType, documentBacking: documentBacking,
							creationDate: creationDate, modificationDate: modificationDate,
							initialPropertyMap: initialPropertyMap)

		// Store
		self.documentInfoMapLock.write() { self.documentInfoMap[documentID] = documentInfo }

		return documentInfo
	}

	//------------------------------------------------------------------------------------------------------------------
	func documentGetInfo(for documentID :String) -> DocumentInfo<T>? {
		// Return document info
		return self.documentInfoMapLock.read() { self.documentInfoMap[documentID] }
	}

	//------------------------------------------------------------------------------------------------------------------
	func documentIterateChanges(
			_ proc
					:(_ documentType :String, _ documentInfoMap :[/* Document ID */ String : DocumentInfo<T>]) throws ->
							Void) rethrows {
		// Setup
		var	map = [/* Document Type */ String : [/* Document ID */ String : DocumentInfo<T>]]()
		self.documentInfoMapLock.read() {
			// Collect info
			self.documentInfoMap.forEach() {
				// Retrieve already collated batch document infos
				if var documentInfoMap = map[$0.value.documentType] {
					// Next document of this type
					map[$0.value.documentType] = nil
					documentInfoMap[$0.key] = $0.value
					map[$0.value.documentType] = documentInfoMap
				} else {
					// First document of this type
					map[$0.value.documentType] = [$0.key : $0.value]
				}
			}
		}

		// Process all document info
		try map.forEach() { try proc($0.key, $0.value) }
	}
}

File: src/main/java/codes/stevobrock/minidocumentstorage/MDSBatchInfo.kt
package codes.stevobrock.minidocumentstorage

import org.w3c.dom.Document
import java.util.*
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write

//----------------------------------------------------------------------------------------------------------------------
class MDSBatchInfo<T : Any> {

	// DocumentInfo
	class DocumentInfo<T : Any> {

		// Properties
				val documentType :String
				val reference :T?
				val creationDate :Date

				var	updatedPropertyMap :HashMap<String, Any>? = null
					private set
				var	removedProperties :HashSet<String>? = null
					private set
				var	modificationDate :Date
					private set
				var	removed = false
					private set

		private	val valueProc :(property :String) -> Any?
		private	val lock = ReentrantReadWriteLock()

		// Lifecycle methods
		//--------------------------------------------------------------------------------------------------------------
		constructor(documentType :String, reference :T?, creationDate :Date, modificationDate :Date,
				valueProc :(property :String) -> Any?) {
			// Store
			this.documentType = documentType
			this.reference = reference
			this.creationDate = creationDate

			this.modificationDate = modificationDate

			this.valueProc = valueProc
		}

		// Instance Methods
		//--------------------------------------------------------------------------------------------------------------
		fun value(property :String) :Any? {
			// Check for document removed
			if (this.removed)
				// Document removed
				return null

			// Check for value
			val	valueProc :() -> Pair<Any?, Boolean>? = {
						// Check the deal
						if (this.removedProperties?.contains(property) ?: false)
							// Property removed
							Pair<Any?, Boolean>(null, true)
						else if (this.updatedPropertyMap?.containsKey(property) ?: false)
							// Property updated
							Pair(this.updatedPropertyMap!![property], false)
						else
							// Property neither removed nor updated
							null
					}
			this.lock.read() { valueProc() }.also() { return it?.first ?: this.valueProc(property) }
		}

		//--------------------------------------------------------------------------------------------------------------
		fun set(property :String, value :Any?) {
			// Write
			this.lock.write() {
				// Check if have value
				if (value != null) {
					// Have value
					if (this.updatedPropertyMap != null)
						// Have updated info
						this.updatedPropertyMap!![property] = value
					else
						// First updated info
						this.updatedPropertyMap = hashMapOf(property to value)

					this.removedProperties?.remove(property)
				} else {
					// Removing value
					this.updatedPropertyMap?.remove(property)

					if (this.removedProperties != null)
						// Have removed properties
						this.removedProperties!!.add(property)
					else
						// First removed property
						this.removedProperties = hashSetOf(property)
				}

				// Modified
				this.modificationDate = Date()
			}
		}

		//--------------------------------------------------------------------------------------------------------------
		fun remove() { this.lock.write() { this.removed = true; this.modificationDate = Date() } }
	}

	// Properties
	private	val documentInfoMap = HashMap</* Document ID */ String, DocumentInfo<T>>()
	private	val documentInfoMapLock = ReentrantReadWriteLock()

	// Instance Methods
	//------------------------------------------------------------------------------------------------------------------
	fun addDocument(documentType :String, documentID :String, reference :T? = null, creationDate :Date,
			modificationDate :Date, valueProc :(property :String) -> Any? = { null }) :DocumentInfo<T> {
		// Setup
		val	documentInfo = DocumentInfo<T>(documentType, reference, creationDate, modificationDate, valueProc)

		// Store
		this.documentInfoMapLock.write() { this.documentInfoMap[documentID] = documentInfo }

		return documentInfo
	}

	//------------------------------------------------------------------------------------------------------------------
	fun documentInfo(documentID :String) :DocumentInfo<T>? {
		// Return document info
		return this.documentInfoMapLock.read() { this.documentInfoMap[documentID] }
	}

	//------------------------------------------------------------------------------------------------------------------
	fun forEach(proc :(documentType :String, documentInfoMap :Map</* Document ID */ String, DocumentInfo<T>>) -> Unit) {
		// Collate
		val	map = HashMap</* Document Type */ String, HashMap</* Document ID */ String, DocumentInfo<T>>>()
		this.documentInfoMapLock.read() {
			// Collect info
			this.documentInfoMap.forEach() {
				// Retrieve already collated batch document infos
				var	documentInfoMap = map[it.value.documentType]
				if (documentInfoMap != null) {
					// Next document of this type
					documentInfoMap[it.key] = it.value
					map[it.value.documentType] = documentInfoMap
				} else
					// First document of this type
					map[it.value.documentType] = hashMapOf(it.key to it.value)
			}
		}

		// Iterate and call proc
		map.forEach() { proc(it.key, it.value) }
	}
}
